esteid/actions.py

```diff
@@ -119,227 +122,429 @@ def do_action(cls, view: "GenericDigitalSignViewMixin", params: dict = None, *,
+                "success": False,
+                "code": "SIGN_SESSION_FAILED",
             }
 
+        signed_digest = sign_session.digest
+        digest_hash_b64 = binascii.b2a_base64(signed_digest).decode()
+
+        update_esteid_session(
+            request,
+            session_id=sign_session.session_id,
+            digest_b64=digest_hash_b64,
+            temp_signature_file=temp_signature_file.name,
+            temp_container_file=temp_container_file.name,
+        )
+
         return {
-            'success': True,
-            'challenge': resp['ChallengeID'],
+            "success": True,
+            "challenge": sign_session.verification_code,
+            "verification_code": sign_session.verification_code,
         }
 
 
 class MobileIdStatusAction(BaseAction):
     @classmethod
-    def do_action(cls, view, action_kwargs):
-        service = view.get_service()
-
-        status_info = service.get_status_info()
+    def do_action(cls, view: "GenericDigitalSignViewMixin", params: dict = None):
 
-        # If error occured
-        if status_info['StatusCode'] not in ['OUTSTANDING_TRANSACTION', 'SIGNATURE', 'REQUEST_OK']:
+        request = view.request
+        session_data = get_esteid_session(request)
+        if not session_data:
             return {
-                'success': False,
-                'code': status_info['StatusCode'],
-                'message': service.MID_STATUS_ERROR_CODES[status_info['StatusCode']],
+                "success": False,
+                "code": "NO_SESSION",
             }
 
-        elif status_info['StatusCode'] == 'OUTSTANDING_TRANSACTION':
+        session_id = session_data["session_id"]
+        signed_digest = session_data["digest_b64"]
+        temp_signature_file = session_data["temp_signature_file"]
+        temp_container_file = session_data["temp_container_file"]
+
+        with open(temp_signature_file, "rb") as f:
+            xml_sig = XmlSignature(f.read())
+
+        # Load a partially prepared BDoc from a tempfile and clean it up
+        container = Container.open(temp_container_file)
+
+        service = TranslatedMobileIDService.get_instance()
+        try:
+            status = service.sign_status(
+                session_id, xml_sig.get_certificate_value(), binascii.a2b_base64(signed_digest)
+            )
+        except ActionNotCompleted:
             return {
-                'success': False,
-                'pending': True,
+                "success": False,
+                "pending": True,
             }
+        except Exception:
+            # NOTE: we could pick some exceptions that don't require cleanup,
+            # but this also requires support from the party that polls this action.
+            # Most likely the whole process would need to be restarted anyway
+            delete_esteid_session(request)
+            raise
+
+        # now we don't need the session anymore
+        delete_esteid_session(request)
```

Comment:
Looks like we can put `delete_esteid_session(request)` into `finally:` and avoid writing it twice. Not sure if it improves readability or makes it worse, though :thinking:

Reply:
No, it unfortunately won't work this way ;-) The point is, when there is an `ActionNotCompleted` exception, we shouldn't delete the session. I will add a comment about that to make it clear.